### ISSUE TITLE
Add package build cache

### ISSUE THREAD

**Issue by rth** [pyodide/pyodide#2179]
We are currently using ccache to cache the compilation of individual objects, which certainly makes the build faster.

However, I wonder if it wouldn't be better to also cache at the package level. Essentially, hash the source tarball, the meta.yaml, patches, emsdk version and env variables passed to the build script. Then take an existing wheel if such hash already exists.

For instance, even with ccache, building scipy still takes very long.

This would allow us to continue having monorepo builds, but waste less resources on rebuilding packages that don't need to be rebuilt.

**Comment by hoodmane** [comment]
Sounds like a good idea to me, particularly for numpy and scipy which are the big bottlenecks. We could start by always rebuilding everything on main and stable but not on other branches?

Is it possible to build things that depend on scipy without rebuilding scipy first? Clearly static libraries are needed at link time, but maybe other things don't actually need to be ordered.

**Comment by rth** [comment]
> We could start by always rebuilding everything on main and stable but not on other branches?

PRs are indeed the main issue.

> Is it possible to build things that depend on scipy without rebuilding scipy first?

That's where it gets a bit more complicated, for packages that are needed at build time such as numpy or scipy and have [`build->skip_host=False`](https://pyodide.org/en/stable/development/meta-yaml.html?highlight=skip_host#build-skip-host) we could try caching both the produced wasm wheel and the installed folder for amd64.
For scipy very few actually need it at build, but for instance scikit-learn does due to the use of Cython BLAS API from scipy.

**Comment by ryanking13** [comment]
Sounds nice to me too. Building packages is really a bottleneck in CI.

> Essentially, hash the source tarball, the meta.yaml, patches, emsdk version and env variables passed to the build script.

What would happen if the static libraries linked to the package has been updated? For example, `zlib` is linked to many packages, and if we add a patch to zlib, we should detect and rebuild the linked packages, if I'm understanding correctly.

**Comment by hoodmane** [comment]
Also, what's the deal with zlib as a library? We already have a copy of zlib in the main module, why do we have to statically link it again into the side modules?

**Comment by rth** [comment]
> What would happen if the static libraries linked to the package has been updated?

Yes, the fact that a dependency was built (as opposed to loaded from cache) should likely trigger the build of all downstream dependencies. We already have [some related logic](https://github.com/pyodide/pyodide/blob/55fbd32ef319c2a2ed5ce297cf4b863a171ac8c4/pyodide-build/pyodide_build/buildpkg.py#L559) so that would probably need updating.

> We already have a copy of zlib in the main module, why do we have to statically link it again into the side modules?

That was part of the long-term effort to move away from Makefiles and build CPython and its dependencies as a regular package (well of course with some special handling due to it being part of the main module). The advantage would be that it would be possible to build CPython dependencies (zlib, sqlite, etc) in parallel, and also generally it makes log reporting, updates etc more systematic. That's what conda does for instance. 

We did move away from Makefiles for C libraries, but kind of stopped there for now. This means we have a duplicate zlib library (it should be pretty fast to build though).
